xds/src/main/java/io/grpc/xds/XdsNameResolver.java

```diff
@@ -1071,21 +876,26 @@ public void run() {
    */
   private static class RoutingConfig {
     private final long fallbackTimeoutNano;
-    private final List<Route> routes;
-    private final boolean applyFaultInjection;
-    @Nullable
-    private final HttpFault faultConfig;
+    // Null if HttpFilter support is enabled but RouterFilter is absent in the filter chain.
```

Comment:
I think this implementation is unnecessarily complicated with many small logic branchings. A very straightforward approach is just having a list of `FilterConfig` (not null, but can be empty). At the beginning processing the chain, you just peek the last entry, if it's not a RouterFilter, append a LameFilter. That's it.

Reply:
>  A very straightforward approach is just having a list of FilterConfig (not null, but can be empty). 

Nullness of the list is needed for whether HttpFilter support enabled or not. That is simple and straightforward. Other tricks can be very misleading. Once HttpFilter support is no longer experimental, the list will be non-null.

Reply:
I mean your code logic contains too many case-specific handlings that can be generalized. That's why you end up having so many `if`s (and `Null if ...`). That introduces many edge cases and is more likely to have bugs, either hidden somewhere already or a small patch could introduce one. And it also makes testing hard/complicated (you would have more code branches need to be covered).

That's why I was saying you could just simply append a lame filter if the last filter is not a router filter, no matter there is one in the middle of the chain or not. And you just have a general code block for processing filters, rather than getting into different codepaths based on if router filter exists or not. The general idea is have a clear/strong specification for each method (AKA do one thing well) with weak precondition/call assumption.

Reply:
The different codepaths is to skip the routing logic if there is no rout filter. This branching is unavoidable even appending a lame filter ahead of time. I do agree that appending a lame filter ahead of time can handle all kinds of filters uniformly and reduce some duplicate code. However, to handle the lame filter in the same way as other filters, it need a synthetic filter name and type_url, which I don't like.

Reply:
Changed as suggested. PTAL again.